File: leetcode/double_pointer/15.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        // 首先，n数之和就是递归的两数之和
        // 两数之和两种思路实现：
        // 1. 排序，双指针向中间走
        // 2. 哈希表找target_sum - current（空间换时间）
        // 但这题有重复项，哈希表会太过复杂，因此采用解法1

        sort(nums.begin(), nums.end());

        vector<vector<int>> all_result;
        int target = 0;     // as specified in description
        vector<int> sub_result(3);
        int sub_target;

        vector<int>::iterator left, right;
        int current_sub_sum;
        for ( auto i = nums.begin(); i != nums.end(); ++i ) {
            sub_result.at(0) = *i;
            sub_target = target - *i;
            left = i+1;             // 每次找sub_result都往右，因为左侧的可能性都遍历完了
            right = nums.end()-1;
            while ( left < right ) {    // 开始在sub范围内缩小范围
                current_sub_sum = *left + *right;
                if ( current_sub_sum > sub_target ) {
                    --right;
                } else if ( current_sub_sum < sub_target ) {
                    ++left;
                } else {
                    sub_result.at(1) = *left;
                    sub_result.at(2) = *right;
                    all_result.push_back(sub_result);
                    // cout << sub_result.at(0) << ' ' << endl;
                    // 跳过和当前sub组合一样的数来避免重复的组合; 由于有排序，所以一样的数会紧贴彼此
                    while ( left < right && *left == sub_result.at(1) ) ++left;
                    while ( left < right && *right == sub_result.at(2) ) --right;
                }
            }
            while ( i+1 != nums.end() && *(i+1) == *i ) ++i;     // same as line 35
        }
        return all_result;
    }
};

File: leetcode/double_pointer/15.ts
function threeSum(nums: number[]): number[][] {
  nums.sort((a, b) => a-b);
  // console.log(nums);
  const n = nums.length;
  const result: [number, number, number][] = [];
  for (let i = 0; i < n; ++i) {
      let k = n-1;    // k必须在第一层重置，虽然i只会往右移动，但理论上只会更小的只是num[j]+num[k]，而不是num[k]，也就是可能出现前几次numij要求的k小于后几次要求的k，因此k需要重置。
      // i 去重
      // i = 0不去重，之后才开始去重
      if ( i > 0 && i < n && nums[i-1] === nums[i] ) continue;
      for ( let j = i+1; j < n; ++j ) {
          // j 去重
          // j-1 = i不去重，之后才开始去重
          if ( j-1 > i && j < n && nums[j-1] === nums[j] ) continue;
          const num12 = nums[i] + nums[j];
          // 移动右指针。k不需要去重，因为j已经去重了、同一个jk组合只会判断一次，且j有可能遍历到和k一样的元素
          while ( k > j && num12 + nums[k] > 0 ) {
              --k;
          }
          if ( k <= j ) break;
          // 检查k停下的是不是目标
          if ( num12 + nums[k] === 0 ) {
              result.push([nums[i], nums[j], nums[k]]);
          }
      }
  }
  return result;
};